### PR TITLE
chore: AGP 8 updates

### LINKED ISSUE #11266: 🐛 Firebase packages don't fully support AGP 8
## 🐛 Bug report

While trying to upload our app to AGP 8, we were getting error logs like this:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':firebase_analytics'.
> defaultConfig contains custom BuildConfig fields, but the feature is disabled.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 914ms
```
The following code snippet fixed the issue for us when building locally. In each plugin's `android/build.gradle` file, you have to add:
```gradle
android {
    buildFeatures {
        buildConfig = true
    }
}
```

### PR DESCRIPTION
**Pull request by Zazo032**
## Description

- Fixes #11266
- Updated Gradle versions

## Related Issues

Closes https://github.com/firebase/flutterfire/issues/11266

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml [hunk 1]
```diff
@@ -12,16 +12,16 @@ environment:
...
+  plugin_platform_interface: ^2.1.4
 
 dev_dependencies:
   firebase_core_platform_interface: ^4.8.0
   flutter_test:
     sdk: flutter
-  mockito: ^5.4.0
-  pigeon: ^9.2.5
+  mockito: ^5.4.2
+  pigeon: ^10.1.3
```
Comment (russellwheatley) [inline packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml:27]: @Lyokone - wondering if this is going to cause an issue?

File: packages/firebase_core/firebase_core_platform_interface/pubspec.yaml [hunk 2]
```diff
@@ -11,15 +11,15 @@ environment:
...
-  meta: ^1.8.0
-  plugin_platform_interface: ^2.1.3
+  meta: ^1.9.1
+  plugin_platform_interface: ^2.1.4
 
 dev_dependencies:
-  mockito: ^5.4.0
-  pigeon: ^9.2.5
+  mockito: ^5.4.2
+  pigeon: ^10.1.3
```
Comment (russellwheatley) [inline packages/firebase_core/firebase_core_platform_interface/pubspec.yaml:25]: @Lyokone - same.

**Review by russellwheatley — CHANGES_REQUESTED** [review]
@Zazo032 - thanks so much for this PR, very much needed 🚀 .

Is it possible you could separate into two PRs? One for android updates and another for pubspec dependency updates? There's a lot of changes and I am concerned about bumping Pigeon up a major version when we've experienced some issues with its integration. The file conflicts will disappear as well because the plugin versions have been bumped since you created this.

for the android updates, I would prefer `compileSdkVersion` and `targetSdkVersion` targeted 33. The AGP version `8.0.2` is only tested up to that level.

FYI; requires using Java 17 which is the LTS version and has been since 2021.

**Comment by Zazo032** [comment]
Thanks for the review! I will remove the `pubspec.yaml` version changes from this PR, and change `compileSdkVersion` and to `targetSdkVersion`.

File: packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml (hunk 1 above)
Comment (Lyokone) [reply packages/firebase_auth/firebase_auth_platform_interface/pubspec.yaml:27]: Should probably be removed 👍 Or we would need to regenerate Pigeon files, but it's probably better to do it in another PR

**Comment by Zazo032** [comment]
@russellwheatley I reverted the `pubspec.yaml`, `compileSdkVersion` and `targetSdkVersion` - feel free to comment if you still find any unwanted change as part of this PR

**Comment by russellwheatley** [comment]
@Zazo032 - great work! There is another change required that I've just noticed. `compileSdkVersion` is deprecated. Could you change them to `compileSdk`, please? 
Screenshot 2023-08-01 at 12 30 28

I would also like any changes to example app code (I noticed changes to app installation and ml model downloader) or pubspec dependencies removed.

Other than those two last bits, this looks almost there. Thanks 😄

**Review by russellwheatley — CHANGES_REQUESTED** [review]
@Zazo032 - great work! There is another change required that I've just noticed. `compileSdkVersion` is deprecated. Could you change them to `compileSdk`, please? 
Screenshot 2023-08-01 at 12 30 28

I would also like any changes to example app code (I noticed changes to app installation and ml model downloader) or pubspec dependencies removed.

Other than those two last bits, this looks almost there. Thanks 😄

**Comment by Zazo032** [comment]
TIL `compileSdkVersion` was deprecated. Do you know why `flutterfire` plugins use `compileSdk 33` but the examples use `compileSdk flutter.compileSdkVersion`? Should I change those example `compileSdk` too?

Just added your comments to the PR. from the example code files, I only left a small typo fix in `firebase_ml_model_downloader` - let me know if I should revert it too

**Comment by russellwheatley** [comment]
@Zazo032 - Yes, please change to 33 for example apps. Typo is fine to stay 👍

**Comment by Zazo032** [comment]
@russellwheatley done!

**Comment by Zazo032** [comment]
@russellwheatley the `e2e / ios (tests) (pull_request)` check is failing due to a timeout, is there something I can do on my side?

**Comment by russellwheatley** [comment]
Don't worry about iOS CI, it doesn't affect this and it is failing across all CI flows at the moment 👍

**Comment by Zazo032** [comment]
@russellwheatley I didn't get that error, what steps should I follow to reproduce it? Looks like it's using 2 different versions, but none of those are defined in the example project

**Comment by vbuberen** [comment]
> Duplicate class kotlin.collections.jdk8.CollectionsJDK8Kt found in modules jetified-kotlin-stdlib-1.8.0 (org.jetbrains.kotlin:kotlin-stdlib:1.8.0) and jetified-kotlin-stdlib-jdk8-1.7.10 (org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.10)

Would like to join the discussion here as recently had similar problems on a few projects while moving from Kotlin `1.7.x` to `1.8.x`. In `1.8` there was this change: https://kotlinlang.org/docs/whatsnew18.html#updated-jvm-compilation-target due to which all projects need to declare stdlib dependency without specifying `jdk7` or `jdk8` from now on (even though, it was possible much earlier).  
The problem is inconsistent as doesn't happen for every single setup. I guess the reason is that some project dependencies still have `jdk7` or `jdk8` specified in their build configs and [the mechanism created](https://kotlinlang.org/docs/whatsnew18.html#usage-of-the-latest-kotlin-stdlib-version-in-transitive-dependencies) by Jetbrains doesn't work in some setups.

In the end, for one big project I came up with this option of Kotlin declaration to be 100% sure that builds won't fail:
```
implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    constraints {
        implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version") {
            because("kotlin-stdlib-jdk7 is now a part of kotlin-stdlib and to not fail builds due duplicate classes")
        }
        implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version") {
            because("kotlin-stdlib-jdk8 is now a part of kotlin-stdlib and to not fail builds due duplicate classes")
        }
    }
```

So I would suggest to either declare Kotlin this way or just drop `-jdk8` which is one of changes in this PR and see if there are issues with build left.

**Comment by russellwheatley** [comment]
@vbuberen - Thank you for the insight, I was able to build once I included the above mentioned code snippet.

@Zazo032 - all the UI packages have now been moved to a separate repo, do you mind updating the PR and removing those changes? 

The next step is to implement the following:

Inside `example/android` `build.gradle`, add the following line if it does not exist:
```gradle
buildscript {
	// The below line
    ext.kotlin_version = '1.9.0'
    repositories {
        google()
        mavenCentral()
    }
```
 Inside the `example/android/app` `build.gradle`, add the following:

```gradle
dependencies {
  // ... whatever dependencies are specified

  implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
  constraints {
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version") {
      because("kotlin-stdlib-jdk7 is now a part of kotlin-stdlib, we constrain to remove duplicate classes")
    }
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version") {
      because("kotlin-stdlib-jdk8 is now a part of kotlin-stdlib, we constrain to remove duplicate classes")
    }
  }
}
```

Thank you for your patience, and sorry for only just returning to this!

**Comment by vbuberen** [comment]
@russellwheatley Would like to get back to the discussion of Kotlin update again as my previous comment was posted based on what I did in apps projects, which were just consumers of plugins. 

Yesterday, I did some additional testing in context of Plus Plugins which I maintain and would like to mention that while the suggested solution to resolve Kotlin versions works well for example apps, when a plugin with Kotlin 1.8.x or 1.9.x is added to some project with older Kotlin version the same issue `duplicate classes` will appear again. So the consuming project has to also have the code snippet I shared above along with new Kotlin version to build Android part. Thus, be aware of this.

Here is my comment on this matter in Plus Plugins: https://github.com/fluttercommunity/plus_plugins/pull/2134#issuecomment-1704047292

**Comment by vbuberen** [comment]
Hey @Zazo032. Are you Ok if I intercept this PR and update it myself to move this topic forward?

**Comment by Zazo032** [comment]
@vbuberen hey! sorry for not following up on this - I'll be out until the second half of October, if you want to prioritise this, feel free to push any changes to the PR (I can check those changes from the phone). If not, I can address your suggestions once I'm back

**Comment by vbuberen** [comment]
Thanks for your reply. Will work on updating this PR. Asked as yesterday met Russel in real life and remembered about this PR.

**Comment by russellwheatley** [comment]
Hey @vbuberen, nice to meet you in Prague 😄. Feel free to take over the PR, I will gladly review! Thanks 👍

**Pull request closed without merging**